AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m

```diff
@@ -0,0 +1,331 @@
+
+#import "MSKeychainUtil.h"
+#import "MSTestFrameworks.h"
+#import "MSTokenExchange.h"
+#import "MSStorageIngestion.h"
+#import "MSTokenResult.h"
+#import "MSDataStoreErrors.h"
+#import "MSTokensResponse.h"
+
+static NSString *const kMSPartition = @"partition";
+static NSString *const kMSToken = @"token";
+static NSString *const kMSStatus = @"status";
+static NSString *const kMSDbName = @"dbName";
+static NSString *const kMSDbAccount = @"dbAccount";
+static NSString *const kMSDbCollectionName = @"dbCollectionName";
+static NSString *const kMSExpiresOn = @"expiresOn";
+static NSString *const kMSTokenResultSucceed = @"Succeed";
+
+static NSString *const expiresOn = @"1999-09-19T11:11:11.111Z";
+static NSString *const token = @"mock-token";
+static NSString *const partition = @"mock-partition";
+
+static NSString *const cachedToken = @"mockCachedToken";
+static NSString *const kMSStorageReadOnlyDbTokenKey = @"MSStorageReadOnlyDbToken";
+static NSString *const kMSStorageUserDbTokenKey = @"MSStorageUserDbToken";
+
+@interface MSTokenExchange (Test)
+
++ (void)removeCachedToken:(NSString *)partitionName;
++ (NSString *)tokenKeyNameForPartition:(NSString *)partitionName;
++ (void)saveToken:(MSTokenResult *)tokenResult;
++ (MSTokenResult *)retrieveCachedToken:(NSString *)partitionName;
+
+@end
+
+@interface MSTokenExchangeTests : XCTestCase
+
+@property(nonatomic) id keychainUtilMock;
+@property(nonatomic) id ingestionStoreMock;
+@property(nonatomic) id sut;
+
+@end
+
+@implementation MSTokenExchangeTests
+
+- (void)setUp {
+  [super setUp];
+  self.sut = OCMClassMock([MSTokenExchange class]);
+  self.keychainUtilMock = OCMClassMock([MSKeychainUtil class]);
+  self.ingestionStoreMock = OCMClassMock([MSStorageIngestion class]);
+}
+
+- (void)tearDown {
+  [super tearDown];
+  [self.sut stopMocking];
+  [self.keychainUtilMock stopMocking];
+  [self.ingestionStoreMock stopMocking];
+}
+
+- (void)testWhenNoCachedTokenNewTokenIsCached {
+  
+  // If
+  NSObject *tokenData = [self getSuccessfulTokenData];
+  NSMutableDictionary *tokenList = [@{kMSTokens : @[ tokenData ]} mutableCopy];
+  NSData *jsonTokenData = [NSJSONSerialization dataWithJSONObject:tokenList
+                                                          options:NSJSONWritingPrettyPrinted
+                                                            error:nil];
+  
+  // Mock HTTP call returning fake token data.
+  OCMStub([self.ingestionStoreMock sendAsync:OCMOCK_ANY completionHandler:OCMOCK_ANY]).andDo(^(NSInvocation *invocation) {
+    MSSendAsyncCompletionHandler completionBlock;
+    [invocation getArgument:&completionBlock atIndex:3];
+    NSHTTPURLResponse *response = [NSHTTPURLResponse new];
+    id mockResponse = OCMPartialMock(response);
+    OCMStub([mockResponse statusCode]).andReturn(200);
+    completionBlock(@"", mockResponse, jsonTokenData, nil);
+  });
+  
+  // Mock returning nil for cached token.
+  NSString *mockTokenName = @"mock-token-name";
```

Comment:
Move `mockTokenName` to a private variable since you declare it ~5 different times

Reply:
Actually, could you just add the below line where you mock the `tokenKeyNameForPartition` method to `setUp`?